File: examples/auction.rb
#!/usr/bin/env ruby

# cf. Scala by Example, Chapter 3

$:.push File.join( File.dirname(__FILE__), "..", "lib" )

require 'dramatis/actor'

class Auction
  include Dramatis::Actor
  def initialize seller, min_bid, closing
    @seller = seller
    @min_bid = min_bid
    @closing = closing

    @time_to_shutdown = 3000
    @bid_increment = 10
    @max_bid = @min_bid - @bid_increment
    @max_bidder = nil

    actor.timeout @closing - Time::now, :timeout
  end

  def inquire
    [ @max_bid, @closing ]
  end

  def offer bid, bidder
    if bid >= @max_bid + @bid_increment
      if @max_bid >= @min_bid
        cast( @max_bidder ).beaten_offer bid
      end
      @max_bid = bid
      @max_bidder = bidder
      :best_offer
    else
      [ :beaten_offer, @max_bid ]
    end
  end

  def timeout
    actor.timeout( @time_to_shtudown  ) { exit }
    actor.become Over.new
  end
end

MIN_BID = 100
CLOSING = Time::now + 4

seller = Dramatis::Actor.new Object.new
auction = Auction.new seller, MIN_BID, CLOSING

class Client
  include Dramatis::Actor
  def initialize id, increment, top, auction
    @id = id
    @increment = increment
    @top = top
    @auction = auction
    @current = 0
    log "started"
    @max = auction.inquire[0]
    log "status #{@max}"
    bid
  end
  def bid
    if @max > @top
      log("too high for me")
    elsif ( @current < @max )
      @current = @max + @increment
      sleep( ( 1 + rand( 1000 ) )/1000.0 )
      answer, max_bid = @auction.offer @current, actor.name
      case answer
      when :best_offer; log("best offer: #{@current}")
      when :beaten_offer; beaten_offer max_bid
      end
    end
  end
  def beaten_offer max_bid
    log("beaten offer: #{max_bid}")
    @max = max_bid
    cast( actor.name ).bid
  end
  def log string
    puts "client #{@id}: #{string}"
  end
end

Client.new 1, 20, 200, auction
Client.new 2, 10, 300, auction

File: examples/auction.py
#!/usr/bin/env python

# cf. Scala by Example, Chapter 3

import time
import random
import inspect
import sys
import os.path

sys.path[0:0] = [ os.path.join( os.path.dirname( inspect.getabsfile( inspect.currentframe() ) ), '..', 'lib' ) ]

from logging import warning

import dramatis

class Auction ( dramatis.Actor ):
    def __init__(self, seller, min_bid, closing):
        self._seller = seller
        self._min_bid = min_bid
        self._closing = closing

        self._time_to_shutdown = 3000
        self._bid_increment = 10
        self._max_bid = self._min_bid - self._bid_increment
        self._max_bidder = None

    def inquire(self):
        return [ self._max_bid, self._closing ]

    def offer(self, bid, bidder):
        if bid >= self._max_bid + self._bid_increment:
            if self._max_bid >= self._min_bid:
                dramatis.release( self._max_bidder ).beaten_offer( bid )
            self._max_bid = bid
            self._max_bidder = bidder
            return [ "best_offer", None ]
        else:
            return [ "beaten_offer", self._max_bid ]

MIN_BID = 100
CLOSING = time.time() + 4

seller = dramatis.Actor( object() )
auction = Auction( seller, MIN_BID, CLOSING )

class Client ( dramatis.Actor ):
    def __init__(self, id, increment, top, auction):
        self._id = id
        self._increment = increment
        self._top = top
        self._auction = auction
        self._current = 0
        self.log( "started" )
        self._max = auction.inquire()[0]
        self.log( "status " + str(self._max) )
        self.bid()

    def bid(self):
        if self._max > self._top:
            self.log("too high for me")
        elif ( self._current < self._max ):
            self._current = self._max + self._increment
            time.sleep( ( 1 + random.randint( 0, 1000 ) )/1000.0 )
            answer, max_bid = self._auction.offer( self._current,
                                                   self.actor.name )
            if answer == "best_offer": self.log("best offer: " + str(self._current))
            elif answer == "beaten_offer": self.beaten_offer( max_bid )

    def beaten_offer(self, max_bid):
        self.log("beaten offer: " + str(max_bid))
        self._max = max_bid
        dramatis.release( self.actor.name ).bid()

    def log(self,string):
        print "client", self._id, ": ", string

Client( 1, 20, 200, auction )
Client( 2, 10, 300, auction )
